src/serde/de.rs

```diff
@@ -0,0 +1,955 @@
+use crate::{IonError, IonResult, IonType, ReaderBuilder, StreamItem, StreamReader, Symbol};
+use chrono::{DateTime, FixedOffset};
+use serde::de;
+use serde::de::{DeserializeSeed, EnumAccess, MapAccess, SeqAccess, Visitor};
+use serde::Deserialize;
+use std::borrow::Cow;
+
+pub fn from_slice<'a, T>(s: &'a [u8]) -> IonResult<T>
+where
+    T: Deserialize<'a>,
+{
+    let mut deserializer = IonDeserializer {
+        reader: ReaderBuilder::new().build(s)?,
+    };
+    // Need to move twice
+    deserializer.reader.next()?; // Nothing header
+
+    let t = T::deserialize(&mut deserializer)?;
+
+    if deserializer.reader.ion_type().is_none() {
+        Ok(t)
+    } else {
+        Err(IonError::DecodingError {
+            description: "expected eof".to_string(),
+        })
+    }
+}
+
+pub fn from_str<'a, T>(s: &'a str) -> IonResult<T>
+where
+    T: Deserialize<'a>,
+{
+    let mut deserializer = IonDeserializer {
+        reader: ReaderBuilder::new().build(s)?,
+    };
+    deserializer.reader.next()?; // Nothing header
+
+    let t = T::deserialize(&mut deserializer)?;
+
+    if deserializer.reader.ion_type().is_none() {
+        Ok(t)
+    } else {
+        Err(IonError::DecodingError {
+            description: "expected eof".to_string(),
+        })
+    }
+}
```

Comment:
Couple of points here.

1. Perhaps these implementations can be shared by a generic implementation that accepts anything implementing `ToIonDataSource`, and each marked with a `#[inline]` hint.
2. Does it really make sense to enforce that `T` deserializes the entirety of the Ion item stream?
4. Related to number 2, an `StreamDeserializer` that implements `Iterator` would be a good addition too, just an idea. A la [`serde_json::de::StreamDeserializer`](https://docs.rs/serde_json/latest/serde_json/de/struct.StreamDeserializer.html).

Reply:
I'll see what i can do to combine them, i'll need to add a flag though for the different number of skips required from binary to string representation (from my testing it looks like i have to do two next calls to get to first object for binary, but only 1 for string). But this brings the problem 2 you mentioned. I'd need to somehow detect if i'm at the start of stream or not

Reply:
Oops ignore the twice thing that was a brain miss

Reply:
Ok 😌 I was going to say, this would mean [`NativeElementIterator` has a bug](https://docs.rs/ion-rs/latest/src/ion_rs/value/native_reader.rs.html#28). Good to hear it's not an issue.

Reply:
Adjusted the methods and also implemented a quick and easy StreamIterator